### ISSUE TITLE
RFC: The Restate widget

### ISSUE THREAD

**Issue by mstijak** [codaxy/cxjs#434]
The `Restate` widget would allow a part of the widget tree to work with a separate (almost completely clean) data store. Data shared beetween the parent store and the child store must be explicitely defined.

**Login Form Example**:
```jsx
<Restate 
  data={{
    userId: bind('$page.userId')
  }}
>
   <form 
       onSubmit={(e, {store}) => { 
          //fetch userId from the server
          store.set('userId', userId); //userId bubbles into the parent store
       }}>
       <TextField value-bind="username" /> //username would not leak into the parent store
       <TextField value-bind="password" /> //password would not leak into the parent store
   </form> 
</Restate>
```

This answers a question about storing of private data for on demand configuration windows that work with deeply nested data, e.g. in a tree or in a particular dashboard widget.

This component would also provide performance benefits as outer parts of the application would not have to be re-rendered on each data change. Similarily, changes on the outside would not go in. 
This approach combines benefits of `IsolatedScope` and `DetachedScope` in a more developer-friendly way.

**Comment by tymarats** [comment]
A controller you might put on `Restate` (or some inner component) would also see these store variables as it would see global ones (which would be overridden in case of name clashes), I assume? The concept might become useful for an array of situations and quite heavily used in the future code, so thumbs up.

**Comment by mstijak** [comment]
Global state would be inaccessible, except for the declared variables. The idea is to get a clean slate and avoid name clashes.

**Comment by tymarats** [comment]
Ah, of course--there's no way of clashing since you're explicitly binding to global variables if you want to access them within `Restate`.

**Comment by mstijak** [comment]
The Restate widget is available in cx@18.6.3. Let's try it internally before finalizing and documenting features.

**Comment by sasatatar** [comment]
First impressions are great! Pretty much any use of prefixes in naming store paths seems like a good use case for `Restate`.
It especially works well in combination with functional components, as it provides them with private state, so no more need for `scope` parameters to isolate internal bindings.

As mentioned in our live discussion, it would be great to implement a flag for logging Restate Store data to console for debugging. 

Also, a more descriptive alias for the developer would be nice, some suggestions are: 
* `LocalState`
* `PrivateState` 
* `IsolatedState` 
* `SandboxedState`
* all the above with the keyword `Store` in place of `State`

**Comment by sasatatar** [comment]
Another benefit of using Restate in functional components is that we no longer have to pay attention if a property is a binding or a literal value, as Restate handles both by default. So no more need for `this.store.set(value.bind, value)` syntax!

In my humble opinion, this will be a new killer feature and a game changer in Cx as it handles many of its short-comings that I was unaware of in the first place! :)

**Comment by sasatatar** [comment]
Perhaps it would be useful if we could initialize ourselves the store for the Restate and pass it to the Restate as a prop:
```
let privateStore = new RestoreStore();
...
<Restate store={privateStore}>
...
</Restate>
```
This way we could have access to the private Store in a controlled manner, but outside the usual Cx context. Currently, it is only possible to pass data to Restate through the data bindings and through `onInit` method if a Controller is used. This offers a lot of possibilities. For example, I could make a method in a parent Controller (outside Restate) that has access to `privateStore` reference, hence can pass data to the Restate component, after the Restate instance has already been initialized.
This feature would be in line with the existing Window behavior, where we can call `window.open();` with a new Store instance: `window.open(new Store());`.

**Comment by sasatatar** [comment]
Yeah, I've figured it after taking a closer look at the code. If needed, it could be done by using an extended RestateStore class and making its set and get methods async, but that may be an anti-pattern.

Another issue we've noticed with `Restate` is that `expressions` and `computables` will cause problems if, for example, they are passed through props to a functional component using Restate:
- If passed as `data` to Restate, Restate will not be able to update their values in parent Store, as they are read-only. 
This can cause hard to detect bugs and unexpected behavior, which is why we should consider outputting a warning or even throwing an error in such cases?
